coba/environments/banditreplay.py

```diff
@@ -0,0 +1,28 @@
+
+from typing import Iterable, Optional
+
+from pandas import DataFrame
+
+from coba.primitives.rewards import SequenceReward
+from coba.environments.primitives import Environment, LoggedInteraction
+from coba.pipes import IterableSource
+from coba.utilities import peek_first
+
+
+
+class BanditReplay(Environment):
+    def __init__(self, df: DataFrame,take: Optional[int] = None):
```

Comment:
Should this have a more generic interface like `Source` or `Sequence` and not take a dependency on pandas DataFrames?

Reply:
I've been thinking about this... What do you think about breaking this functionality into two pieces?

1. `DataFrameSource`
2. `SimpleEnvironment`

Then in Environments there would be something like:

```
class Environments:
    def from_data(self, data: 'DataFrame'): #note the quotes, this should let us document without dependencies
        return Environments(Pipes.join(DataFrameSource(data),SimpleEnvironment()))
```

Then in the future we can add support for new data formats by simply adding more sources.

`DataFrameSource` would be super super simple. Something like:

```
class DataFrameSource:

    def ___init__(self, dataframe) -> None:
        self._df = df

    def read(self) -> Iterable[<something generic like dict>]:
       return self._df.iterrows()
```

And then `SimpleEnvironment` would be something like:

```
class SimpleEnvironment:

def filter(self, data: Iterable[<something generic like dict>]) -> Iterable[Interaction]:
    for item in data:
      # code here to turn data into interactions
```

Reply:
(Oh I'm also thinking SimpleEnvironment would be "generic" in the since that it could return both SimulatedInteraction and LoggedInteraction depending on what fields items in data have. So if data is only context,action,reward,probability we return a LoggedInteraction. If an item in data has context,actions,rewards we return SimulatedInteraction. Or something along those lines.)